### PR TITLE
Add experiment type

### PR DESCRIPTION
**Pull request by toastisme**
This adds `Experiment.get_type()`, which returns an  `ExperimentType` enum. The idea is to replace `Experiment.is_still()/Experiment.is_sequence()` with something more flexible.

There is an argument that this goes a bit against the dials as a toolkit ethos,  where you should be able to mix pipelines of algorithms. However, the workflow for stills and rotational experiments are distinct enough that a quick way of identifying them is useful (`is_still()`), and time of flight experiments are similar in being quite separate. 

Experiments do not have a static enum applied to their type. Instead it's calculated on the fly to ensure it's always consistent and defined in only one place.

### CODE REVIEW

**Comment by toastisme** [comment]
@phyy-nx @dwpaley do you guys have any opinions on this?

File: src/dxtbx/model/__init__.py [hunk 1]
```diff
@@ -599,11 +602,20 @@ def imagesets(self):
 
     def all_stills(self):
         """Check if all the experiments are stills"""
-        return all(exp.is_still() for exp in self)
+        return all(exp.get_type() == ExperimentType.still for exp in self)
 
     def all_sequences(self):
         """Check if all the experiments are from sequences"""
-        return all(exp.is_sequence() for exp in self)
+        warnings.warn("all_sequences() is deprecated. Use all_rotations() instead")
```
Comment (graeme-winter) [inline src/dxtbx/model/__init__.py:609]: A still `sequence` is meaningful e.g. grid scan

File: src/dxtbx/model/boost_python/experiment.cc [hunk 2]
```diff
@@ -74,6 +74,11 @@ namespace dxtbx { namespace model { namespace boost_python {
   };
 
   void export_experiment() {
+    enum_<ExperimentType>("ExperimentType")
+      .value("still", still)
```
Comment (graeme-winter) [inline src/dxtbx/model/boost_python/experiment.cc:78]: What about a sequence of stills?

File: src/dxtbx/model/experiment.h [hunk 3]
```diff
@@ -28,6 +28,8 @@
 
 namespace dxtbx { namespace model {
 
+  enum ExperimentType { rotation = 1, still = 2, tof = 3 };
```
Comment (graeme-winter) [inline src/dxtbx/model/experiment.h:31]: Enumerations are usually `ENUMERATED`?

File: src/dxtbx/model/boost_python/experiment.cc (hunk 2 above)
Comment (graeme-winter) [reply src/dxtbx/model/boost_python/experiment.cc:78]: I appreciate that the behaviour of the current system is not well defined?

**Review by graeme-winter — COMMENTED** [review]
If we can resolve the definition of sequences of stills, I am happy

**Comment by ndevenish** [comment]
Discussion for core:
- Make is_still/is_sequence should not be deprecated (here)
- is_still/is_sequence should directly call get_type instead of calculating themselves
- CAPITAL_ENUMS

**Pull request merged**